lib/Parser.js

```diff
@@ -1006,47 +1006,103 @@ class Parser extends Tapable {
 
 	prewalkForStatement(statement) {
 		if (statement.init) {
-			if (statement.init.type === "VariableDeclaration")
+			if (
+				statement.init.type === "VariableDeclaration" &&
+				statement.init.kind === "var"
+			)
 				this.prewalkStatement(statement.init);
 		}
 		this.prewalkStatement(statement.body);
 	}
 
 	walkForStatement(statement) {
 		if (statement.init) {
-			if (statement.init.type === "VariableDeclaration")
-				this.walkStatement(statement.init);
-			else this.walkExpression(statement.init);
+			if (statement.init.type === "VariableDeclaration") {
+				if (statement.init.kind === "var")
+					this.walkVariableDeclaration(statement.init);
+				else {
+					this.inBlockScope(statement.init, () => {
```

Comment:
While walking, if variable is block scoped, then execute the whole for loop along with declaration in a new block, else execute normally.